### ISSUE TITLE
Feature request: ability to pass training_hooks and others through keras.estimator.model_to_estimator

### ISSUE THREAD

**Issue by zmjjmz** [tensorflow/tensorflow#18247]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: I'm trying to
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: n/a
- **TensorFlow installed from (source or binary)**: pip
- **TensorFlow version (use command below)**: n/a
- **Python version**: n/a
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a 
- **Exact command to reproduce**: n/a

### Describe the problem

When I go to create an Estimator from a Keras model, I'm unable to pass training hooks, evaluation hooks, and other fun things (like scaffolds) since the `_create_keras_model_fn` call doesn't take any of that stuff in. 

Could it simply just take these as arbitrary `**kwargs` and pass them through to the `EstimatorSpec` initializer?

### Source code / logs

Here's the relevant code: https://github.com/tensorflow/tensorflow/blob/r1.7/tensorflow/python/keras/_impl/keras/estimator.py#L364

Thanks!

**Comment by asimshankar** [comment]
@ispirmustafa @fchollet : Any suggestions here?

**Comment by ispirmustafa** [comment]
Hi @zmjjmz 
Could you please elaborate on use cases?

My current thinking is as follows, it may change based on your use case:

model_to_estimator recreates the same model within an estimator. Since that behavior does not need a hook or Scaffold, we should not add such an argument to model_to_estimator.

you can add hooks when you call the estimator.train(..., hooks=) or evaluate

If you need to play with hooks, scaffold and other estimator spec parameter I would recommend writing your model_fn which should not be that complicated. Here is the pseudo code:
def model_fn(features, labels, mode):
  model = create_your_keras_model
  logits = model(features)
  head = pick_a_reasonable_head_from_tf.contrib.estimator
  spec = head.create_estimator_spec(...)
  return spec._replace(scaffold=..., training_hooks=...)

**Comment by zmjjmz** [comment]
Actually this is largely due to trying to use keras with Amazon SageMaker. As far as I can tell, I don't have access to the actual train call when using the SageMaker API (afaict), and as such am putting hooks in the `EstimatorSpec` initializer. 

My current (broken) keras model function as such looks a bit like this:

```
def create_basic_keras_model_fn(model_class):                                                                                                       
    def model_fn(features, labels, mode, params):                                                          
        inp_placeholder = features[inp_name]                                                           
        model_artifacts = serialization_utils.load_cPkl(artifacts_location)                                
        model_builder = model_class(model_artifacts, **model_parameters)                                   
        model_builder.build_model(inp_placeholder)                                                         
        tensors = model_builder.give_outputs()                                                             
        model = keras.models.Model(inputs=[inp_placeholder], outputs=tensors.values())                     
                                                                                                                                                                                            
        model.compile(**compile_spec)                                                                      
                                                                                                           
        # call _create_keras_model_fn                                                                      
        # we'll pass our features labels mode etc. thru to the resulting model_fn to get our EstimatorSpec 
        # this hurts my head but makes sense                                                               
        keras_model_fn = _create_keras_model_fn(model, custom_objects=ml_utils.KERAS_CUSTOM_LAYERS)        
        return keras_model_fn(features, labels, mode)                                                      
```

As such I don't really have direct access to the EstimatorSpec, so I can't call that `_replace` method there.

> Since that behavior does not need a hook or Scaffold

Can I add hooks and such to the keras model itself prior to the conversion? That would definitely solve my issue, however I don't know if it's possible. I know I can add Callbacks, but I don't think those are easily converted to hooks (but I really wish they were).

Also I'm not sure what you mean by `head = pick_a_reasonable_head_from_tf.contrib.estimator` -- is that this? https://www.tensorflow.org/api_docs/python/tf/contrib/learn/Head

**Comment by ispirmustafa** [comment]
Great for giving us background.
one solution, would be asking sage maker a way to define hooks if you're using estimators.

another solution is using model_fn extension as follows:
`est = model_to_estimator(...)
def new_model_fn(features, labels, mode, config):
  spec = est.model_fn(features, labels, mode, config)
  return spec._replace(...)
`  
head: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/estimator/python/estimator/head.py

for example `tf.contrib.estimator.multi_head`
